src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that stores and shows images and comments. */
+@WebServlet("/search")
+public class SearchServlet extends HttpServlet {
+
+  @Override
+  /* Show courses. */
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    List<Course> courses = getHelper(request);
+
+    String coursesJson = new Gson().toJson(courses);
+    response.setContentType("application/json;");
+    response.getWriter().println(coursesJson);
+  }
+
+  /* Create list of courses given request. Used for testing. */
+  public List<Course> getHelper(HttpServletRequest request) {
+    List<Filter> filters = getFilters(request);
+    List<Entity> results = getResults(filters);
+    List<Course> courses = getCourses(results);
+    return courses;
+  }
+
+  /* Create list of filters given parameters specified in request. */
+  public List<Filter> getFilters(HttpServletRequest request) {
```

Comment:
Looks like this should be private, same for several other methods here

Reply:
When I try to test any function that calls these helper functions (like doGet or getHelper), I get an error if I don't make them public because they're not accessible - am I calling these functions incorrectly or anything like that?

Reply:
If you just call `doGet` directly in your test you get an error if, say, `getFilters` isn't public?

Reply:
yes - it says something along the lines of "getFilters is private so it can't be accessed outside of this file"

Reply:
That seems to imply that you're trying to call `getFilters` directly somewhere, is that the case?

Reply:
No :/

Reply:
Which line is the error from?

Reply:
For some reason, I think it got fixed now, so that's good!